Add final? 0.7.0 highlights and group by topic (#944)
napari/docs: docs/release/release_0_7_0.md

Reviewer: I would probably hold off on trumpeting this like this? 
It's still pretty bad, so people will still have a poor experience if they try IMO.
I have a more comprehensive PR to better address the underlying performance issue:
https://github.com/napari/napari/pull/8636

Reviewer 2: (Sorry about that waiting PR, I've been looking into it often but it's always a bit overwhelming and long to review :sweat_smile: )

Reviewer: No worries!

Author: @psobolewskiPhD I've reworded a little and flagged that we're still working on it - happy with that?

channel](https://napari.zulipchat.com/#narrow/channel/309872-plugins) or by
coming to one of our [community meetings](meeting-schedule).

> #### Negative axis labels? A real positive

Reviewer 3: 😂

If you've ever loaded data of mixed dimensionality in napari, like a TYX volume
alongside a YX segmentation, you may have noticed the default axis labels didn't
quite line up:

| axes   | 0 | 1 | 2 |
|--------|---|---|---|
| volume | 0 | 1 | 2 |
| segmt  |   | 0 | 1 |

That's because napari used 0-based indexing for its viewer axis labels, which breaks
down when layers have different numbers of dimensions. With
[#8565](https://github.com/napari/napari/pull/8565),

> viewer axis labels now use negative indexing by default, just like Python's own indexing

Reviewer 3: ```suggestion
viewer axis labels now use negative indexing by default, similar to Python's own indexing
```

(or: combining Python's negative indexing semantics with NumPy's array broadcasting semantics. I feel like "just like Python" kinda overstates the similarity...)

a long-standing bug where axis labels could end up duplicated when mixing layers of
different dimensionality ([#6569](https://github.com/napari/napari/issues/6569)).

You'll notice this change in the dims slider labels, the axis overlay, and the dims
popup widget. If you already label your axes with your own names (e.g. `z`, `y`, `x`),
nothing's changed. For everyone else, we have consistency at last!

#### What's in an angle? The truth! Fixed camera angles 🎥

If you've ever set up the camera to take that perfect publication-worthy photo of
your data (and taken the time to query the camera angles), you may have noticed they seemed... off.
That's because they were! Very... off. This was due to a long-standing bug in how we calculated our
camera angles, fueled in part by some arcane vispy axis-swapping tomfoolery, and in part by napari's
starting position of `viewer.camera.angles = (0, 0, 90)`.

Good news! With [#8281](https://github.com/napari/napari/pull/8281), angles make sense again. The default camera angles are `(0, 0, 0)`, and they
move intuitively -- so `viewer.camera.angles = (0, 0, 10)` actually represents a 10 degree

> rotation around the 0th dimension. What a time to be alive!

Reviewer 3: 0th or -1st? 🤔 

Just checked, -1st:

```suggestion
rotation around the last ([-1st](axis-names)) dimension. What a time to be alive!
```

Reviewer 3: Oh, seeing below I think you actually meant to write (10, 0, 0), and 0th is right. Anyway, fix one or the other thing. 😅

Reviewer 2: I tguess we should callit -3rd now? xD

Old versions of napari:

![Image showing an old version of a napari viewer with a layer opened and its camera angle (10, 0, 0) displayed in the console.](https://github.com/user-attachments/assets/9ae2040c-36f7-4c4c-8ef8-140202d7ccda)

New and sane:

![Image showing the 0.7.0 napari viewer with a layer opened and its camera angle (10, 0, 0) displayed in the console. The layer is rotated 10 degrees in its first dimension](https://github.com/user-attachments/assets/6b972b46-5c3c-439a-8b0a-fe8a293224e5)

All rotations are now right-handed (counterclockwise when the axis points towards the viewer),
with automatic sign-flipping for flipped camera views. We've also removed the unwieldy to type

> (and confusing to reason about) `quaternion2euler_degrees` in favour of scipy's `Rotation` class.

Reviewer 3: removed from where? Maybe remove this sentence? I doubt anyone (?) is going to know what you're talking about here...

Reviewer 2: yeah agree, this is likely unkown to most

Now for the bad news... After many (and we do [mean](https://github.com/napari/napari/pull/8537)
[**many**](https://github.com/napari/napari/pull/8557)) attempts, we realized we couldn't
provide legacy conversion functions to get you to and from the original camera angles. Therefore,
this is a **breaking change**.

If you had scripts or notebooks setting up angles for screenshots, or if you've got workshop
materials or tutorials with preset angles, they'll need to be updated. Any existing code
using `viewer.camera.angles = (z, y, x)` will now produce a different view than before.

### New features & widgets

#### What's my metadata? Where's my metadata? `napari-metadata` to the rescue

> With a lot of work from our community contributor, Carlos Mario Rodriguez Reza (@carlosmariorr), and

Reviewer 3: these should be manual links, e.g.:

```suggestion
With a lot of work from our community contributor, Carlos Mario Rodriguez Reza ([@carlosmariorr](https://github.com/carlosmariorr)), and
```

because they are only autolinked on GH, not on the website or any other renderer.

Reviewer 3: (or even `[**@carlosmariorr**]`)

![GIF displaying the highlights on the Shapes and Points new layer buttons when one or more layers are selected in the layerlist](https://github.com/user-attachments/assets/dba88d45-baa9-47df-80e9-5c7b1f2a711d)

PS -- You can now also create these new layers from the `File -> New Layer` menu!

#### Better text overalys 🔡

With [#8236](https://github.com/napari/napari/pull/8236), we've not only refactored text overlays
so they're easier to implement, but we've also introduced two new long-requested overlays:
the layer name overlay, and an overlay for the current slice. Together, they make generating
publication-ready figures much easier!

![Image showing the napari viewer with two layers in grid mode. Each layer has its name displayed in the top left, and the current slice displayed in the bottom right.](https://github.com/user-attachments/assets/3c96b38d-44c1-432b-b294-aa9c0934a553)

Try it yourself:

```python
import napari
v = napari.Viewer()
v.grid.enabled = True
ll = v.open_sample('napari', 'cells3d')
for l in ll:
    l.name_overlay.visible = True
v.scale_bar.visible = True
v.scale_bar.gridded = True
v._overlays['current_slice'].visible = True

> v._overlays['current_slice'].gridded = True

Reviewer 2: Maybe note that this is not public for now cause we'r enot sure on the api.

Author: @brisvag added a note in the latest commit.

v.dims.axis_labels = ['z', 'y', 'x']
```

### Rendering & display

#### More pixels to play with - texture tiling

Ever loaded a large 2D image in napari just to zoom in and find it's blurrier

> than a JPEG from the year 2000? That's no longer the case!

Reviewer: I know this is mostly for fun, but I don't love conflating resolution with a compression codec, one that could be lossless. It's already something a lot of people confuse...
Can we use a different phrasing?

Author: I've reworded while trying to keep it fun - let me know if you'd like to workshop it a little still

Courtesy of our community contributor, Guillaume Witz (@guiwitz), and his PR for
texture tiling ([PR #8395](https://github.com/napari/napari/pull/8395)) 2D
images that exceed OpenGL's maximum texture size will be split into multiple
tiles, each small enough to fit on the GPU.

![Image with a screenshot of napari 0.6.6 on the left and napari 0.7.0 on the right displaying a DeCAM image of the Milky Way. The image on the left is pixelated, while the image on the right is displayed at full resolution.](https://github.com/user-attachments/assets/d0a115a8-49d5-432c-b561-f29fe9ac8116)

#### Points - any size you like 🟣

> On MacOS, the points layer has never been able to reach its full potential, as OpenGL

Reviewer: ```suggestion
On macOS, the points layer has never been able to reach its full potential, as OpenGL
```

drivers limit the size of an individual marker to a certain number of screen pixels.

With [#8552](https://github.com/napari/napari/pull/8552) and the release of `vispy v0.16`,
this long-standing issue has finally been resolved. Across all operating systems, you can
make your points as big as you want!

> This change has also propagated to the zoom behaviour on MacOS -- points now zoom

Reviewer: ```suggestion
This change has also propagated to the zoom behaviour on macOS -- points now zoom
```

the layer name overlay, and an overlay for the current slice. Together, they make generating
publication-ready figures much easier!

![Image showing the napari viewer with two layers in grid mode. Each layer has its name displayed in the top left, and the current slice displayed in the bottom right.](https://github.com/user-attachments/assets/3c96b38d-44c1-432b-b294-aa9c0934a553)

Try it yourself:

```python
import napari
v = napari.Viewer()
v.grid.enabled = True
ll = v.open_sample('napari', 'cells3d')
for l in ll:
    l.name_overlay.visible = True
v.scale_bar.visible = True
v.scale_bar.gridded = True
v._overlays['current_slice'].visible = True
v._overlays['current_slice'].gridded = True
v.dims.axis_labels = ['z', 'y', 'x']
```

**Note**: the `v._overlays` attribute is still private as we're working out the best API.

### Rendering & display

#### More pixels to play with - texture tiling

Ever loaded a large 2D image in napari just to zoom in and feel like you're not
really getting a lot of bang for your pixel bucks? That's because we were
downsampling images that were too large to send the whole thing to the GPU.

> Courtesy of our community contributor, Guillaume Witz (@guiwitz), and his PR for

Reviewer 3: I find "our community contributor" a bit overused. I prefer simple acknowledgements, e.g., "Thanks to Guillaume Witz,"

I'd also add a line below like, "If you want to try it out, you could do worse than [this gigapixel astronomy image from the NSF NOIRLab](https://noirlab.edu/public/images/noirlab2027a/)"

Author: > I find "our community contributor" a bit overused. 

Lucky us hehe!

tiles, each small enough to fit on the GPU.

![Image with a screenshot of napari 0.6.6 on the left and napari 0.7.0 on the right displaying a DeCAM image of the Milky Way. The image on the left is pixelated, while the image on the right is displayed at full resolution.](https://github.com/user-attachments/assets/d0a115a8-49d5-432c-b561-f29fe9ac8116)

#### Points - any size you like 🟣

On macOS, the points layer has never been able to reach its full potential, as OpenGL
drivers limit the size of an individual marker to a certain number of screen pixels.

With [#8552](https://github.com/napari/napari/pull/8552) and the release of `vispy v0.16`,
this long-standing issue has finally been resolved. Across all operating systems, you can
make your points as big as you want!

This change has also propagated to the zoom behaviour on macOS -- points now zoom
proportionally to the data, rather than staying the same size in screen pixels.

Here's the behaviour pre 0.7.0:

![Video with a points layer on a grid of white squares. When zooming, the points stay the same size in screen pixels.](../_static/images/points_zoom_066.webm)

And now:

> ![Video with a points layer on a grid of white squares. When zooming, the points scale proportionally to the data.](../_static/images/points_zoom_070.webm)

Reviewer 3: ah with webm you can't just use markdown, this is the syntax:

https://github.com/napari/docs/blob/a2feafd21cc696c2058bdea6f7d526c5dcd8f31f/docs/release/release_0_5_5.md?plain=1#L38-L49

(and yes, you refer to the mp4 without seeing it, because it is automatically generated during docs build, thanks to @melissawm! 🤩 I'm not sure about the image fallback, that might need to be done manually...)

Reviewer 3: Result: https://napari.org/dev/release/release_0_5_5.html

Reviewer 4: Yes the image fallback is a screenshot I do take manually - let me know if you need help!

Reviewer 5: I'll bring it up because I always do, but Github kind of competes with how we build our docs
Looking at the release notes on Github (which I think newer Github folks like me often do) you would just see this html, even though it shows in the docs. I used github content as gifs for my release notes, but I know that Juan always didn't like that they weren't videos (and I agree).

https://github.com/DragaDoncila/docs/blob/b27ec8b071d255c0c2607a9d1e1708b21a6b1158/docs/release/release_0_7_0.md#points---any-size-you-like-

Author: yeah I know @TimMonko I usually have done gifs, but Juan doesn't like gifs and they are kinda huge tbf. Also limited to 10MB via gh-lfs, so I opted for a video here. I do think I prefer having the docs rendered nicely with a high-res video to having the github release notes fully functional (in a world where we can't have both), so I'm going to leave this as is.

would lock up the viewer entirely. Not anymore!

0.7.0 brings a flurry of performance improvements:

- Box selection now uses bounding boxes and vectorized intersection tests,
  delivering a more than 100x speedup ([#8378](https://github.com/napari/napari/pull/8378)).
  Selecting 10,000 shapes goes from >50s to ~0.3s.
- Batch deletion replaces one-by-one removal for another 100x speedup
  ([#8375](https://github.com/napari/napari/pull/8375))! Deleting 50,000 shapes
  now takes under half a second.
- Outline computation is batched and cached, so zooming and panning with
  selected shapes no longer blocks
  ([#8403](https://github.com/napari/napari/pull/8403),
  [#8536](https://github.com/napari/napari/pull/8536)).
- Highlight updates are throttled for large layers, enabling smooth zoom
  even with 200,000+ shapes ([#8404](https://github.com/napari/napari/pull/8404)).
- Mode switching no longer triggers unnecessary redraws, giving another
  ~3x speedup when many shapes are selected
  ([#8551](https://github.com/napari/napari/pull/8551)).

There's still more to do (drawing and drag-moving large selections remain

> slow), but the days of the viewer locking up on a big shapes layer are over.

Reviewer: again, let's not oversell. My experience is that it's best to *undersell* and *over deliver*
```suggestion
Beware: there's still more to do, because drawing and drag-moving large selections remain
slow!
```
